File: src/main/scala/com/xebia/functional/vectorstores/postgres/PGVectorStore.scala
package com.xebia.functional.vectorstores.postgres

import java.util.UUID

import cats.effect.Resource
import cats.effect.Sync
import cats.syntax.all.*

import com.xebia.functional.config.DBConfig
import com.xebia.functional.domain.Document
import com.xebia.functional.embeddings.Embeddings
import com.xebia.functional.embeddings.models.Embedding
import com.xebia.functional.embeddings.openai.models.RequestConfig
import com.xebia.functional.vectorstores.VectorStore
import com.xebia.functional.vectorstores.models.DocumentVectorId
import com.xebia.functional.vectorstores.postgres.PGSql.*
import doobie.ConnectionIO
import doobie.hikari.HikariTransactor
import doobie.implicits._
import doobie.util.transactor.Transactor

class PGVectorStore[F[_]: Sync](
    dbConfig: DBConfig,
    embeddings: Embeddings[F],
    collectionName: String,
    distanceStrategy: PGDistanceStrategy,
    preDeleteCollection: Boolean,
    transactor: Resource[F, Transactor[F]],
    requestConfig: RequestConfig,
    chunckSize: Option[Int]
) extends VectorStore[F]:
  def delCollection: ConnectionIO[Unit] =
    preDeleteCollection match
      case false => Sync[ConnectionIO].unit
      case true =>
        for
          collection <- getCollection(collectionName).option
          c <- collection.liftTo[ConnectionIO](PGErrors.CollectionNotFound(collectionName))
          _ <- deleteCollectionDocs(c.uuid).run
          _ <- deleteCollection(c.uuid).run
        yield ()

  def initialDbSetup(): F[Unit] =
    transactor
      .use { xa =>
        (for {
          _ <- addVectorExtension.run
          _ <- createCollectionsTable.run
          _ <- createEmbeddingTable(dbConfig.vectorSize).run
          _ <- delCollection
        } yield ()).transact(xa)
      }.adaptErr { case e: Throwable => PGErrors.DatabaseSetupError(e.getMessage) }

  def createCollection: F[Int] =
    transactor.use { xa =>
      memeid4s.UUID.V1.next
        .asJava()
        .pure.flatMap { uuid =>
          addNewCollection(uuid, collectionName).run.transact(xa)
        }
    }

  def addTexts(texts: List[String]): F[List[DocumentVectorId]] =
    transactor.use { xa =>
      for
        embs <- embeddings.embedDocuments(texts, chunckSize, requestConfig)
        pairs = texts zip embs
        us <-
          (for
            collection <- getCollection(collectionName).option
            c <- collection.liftTo[ConnectionIO](PGErrors.CollectionNotFound(collectionName))
            uuids <-
              pairs.traverse((t, e) =>
                memeid4s.UUID.V1.next
                  .asJava()
                  .pure[ConnectionIO]
                  .flatTap(u => addNewText(u, c.uuid, t, e).run)
              )
          yield (uuids.map(DocumentVectorId(_)))).transact(xa)
      yield us
    }

  def addDocuments(documents: List[Document]): F[List[DocumentVectorId]] =
    addTexts(documents.map(_.content))

  def similaritySearch(query: String, k: Int): F[List[Document]] =
    transactor.use { xa =>
      for
        qEmb <- embeddings.embedQuery(query, requestConfig)
        docs <- qEmb match
          case e +: es => searchSimilarDocument(e, distanceStrategy, k).to[List].transact(xa)
          case _ => Sync[F].raiseError(PGErrors.EmbeddingNotGenerated(query))
      yield docs
    }

  def similaritySearchByVector(embedding: Embedding, k: Int): F[List[Document]] =
    transactor.use { xa =>
      searchSimilarDocument(embedding, distanceStrategy, k).to[List].transact(xa)
    }

object PGVectorStore:
  def make[F[_]: Sync](
      config: DBConfig,
      embeddings: Embeddings[F],
      collectionName: String,
      strategy: PGDistanceStrategy,
      preDeleteCollection: Boolean,
      transactor: Resource[F, Transactor[F]],
      requestConfig: RequestConfig,
      chunkSize: Option[Int]
  ): PGVectorStore[F] =
    new PGVectorStore[F](config, embeddings, collectionName, strategy, preDeleteCollection, transactor, requestConfig, chunkSize)

  def fromTexts[F[_]: Sync](
      texts: List[String],
      requestConfig: RequestConfig,
      config: DBConfig,
      embeddings: Embeddings[F],
      collectionName: String,
      strategy: PGDistanceStrategy,
      preDeleteCollection: Boolean,
      transactor: Resource[F, Transactor[F]],
      chunkSize: Option[Int]
  ): F[PGVectorStore[F]] =
    for
      store <- Sync[F].pure(make[F](config, embeddings, collectionName, strategy, preDeleteCollection, transactor, requestConfig, chunkSize))
      _ <- store.initialDbSetup()
      _ <- store.createCollection
      _ <- store.addTexts(texts)
    yield (store)

  def fromDocuments[F[_]: Sync](
      documents: List[Document],
      requestConfig: RequestConfig,
      config: DBConfig,
      embeddings: Embeddings[F],
      collectionName: String,
      strategy: PGDistanceStrategy,
      preDeleteCollection: Boolean,
      transactor: Resource[F, Transactor[F]],
      chunkSize: Option[Int]
  ): F[PGVectorStore[F]] =
    for
      store <- Sync[F].pure(make[F](config, embeddings, collectionName, strategy, preDeleteCollection, transactor, requestConfig, chunkSize))
      _ <- store.initialDbSetup()
      _ <- store.createCollection
      _ <- store.addDocuments(documents)
    yield (store)

File: src/jvmMain/kotlin/com/xebia/functional/PGVectorStore.kt
package com.xebia.functional

import com.xebia.functional.embeddings.Embedding
import com.xebia.functional.embeddings.Embeddings
import com.xebia.functional.llm.openai.RequestConfig
import com.xebia.functional.vectorstores.DocumentVectorId
import com.xebia.functional.vectorstores.PGCollection
import com.xebia.functional.vectorstores.PGDistanceStrategy
import com.xebia.functional.vectorstores.VectorStore
import com.xebia.functional.vectorstores.addNewCollection
import com.xebia.functional.vectorstores.addNewText
import com.xebia.functional.vectorstores.addVectorExtension
import com.xebia.functional.vectorstores.createCollectionsTable
import com.xebia.functional.vectorstores.createEmbeddingTable
import com.xebia.functional.vectorstores.deleteCollection
import com.xebia.functional.vectorstores.deleteCollectionDocs
import com.xebia.functional.vectorstores.getCollection
import com.xebia.functional.vectorstores.searchSimilarDocument
import javax.sql.DataSource
import kotlinx.uuid.UUID
import kotlinx.uuid.generateUUID

class PGVectorStore(
  private val vectorSize: Int,
  private val dataSource: DataSource,
  private val embeddings: Embeddings,
  private val collectionName: String,
  private val distanceStrategy: PGDistanceStrategy,
  private val preDeleteCollection: Boolean,
  private val requestConfig: RequestConfig,
  private val chunckSize: Int?
) : VectorStore {

  suspend fun JDBCSyntax.getCollection(collectionName: String): PGCollection =
    queryOneOrNull(getCollection,
      { bind(collectionName) }
    ) { PGCollection(UUID(string()), string()) }
      ?: throw IllegalStateException("Collection '$collectionName' not found")

  suspend fun JDBCSyntax.deleteCollection() {
    if (preDeleteCollection) {
      val collection = getCollection(collectionName)
      update(deleteCollectionDocs) { bind(collection.uuid.toString()) }
      update(deleteCollection) { bind(collection.uuid.toString()) }
    }
  }

  suspend fun initialDbSetup(): Unit = dataSource.connection {
    update(addVectorExtension)
    update(createCollectionsTable)
    update(createEmbeddingTable(vectorSize))
    deleteCollection()
  }

  suspend fun createCollection(): Unit = dataSource.connection {
    val xa = UUID.generateUUID()
    update(addNewCollection) { bind(xa.toString()); bind(collectionName) }
  }

  override suspend fun addTexts(texts: List<String>): List<DocumentVectorId> = dataSource.connection {
    val embeddings = embeddings.embedDocuments(texts, chunckSize, requestConfig)
    val collection = getCollection(collectionName)
    texts.zip(embeddings) { text, embedding ->
      val uuid = UUID.generateUUID()
      update(addNewText) {
        bind(uuid.toString())
        bind(collection.uuid.toString())
        bind(embedding.data.toString())
        bind(text)
      }
      DocumentVectorId(uuid)
    }
  }

  override suspend fun addDocuments(documents: List<Document>): List<DocumentVectorId> =
    addTexts(documents.map(Document::content))

  override suspend fun similaritySearch(query: String, limit: Int): List<Document> = dataSource.connection {
    val embeddings = embeddings.embedQuery(query, requestConfig).ifEmpty { throw IllegalStateException("Embedding for text: '$query', has not been properly generated") }
    val collection = getCollection(collectionName)
    queryAsList(searchSimilarDocument(distanceStrategy), {
      bind(collection.uuid.toString())
      bind(embeddings[0].data.toString())
      bind(limit)
    }) { Document(string()) }
  }

  override suspend fun similaritySearchByVector(embedding: Embedding, limit: Int): List<Document> =
    dataSource.connection {
      val collection = getCollection(collectionName)
      queryAsList(searchSimilarDocument(distanceStrategy), {
        bind(collection.uuid.toString())
        bind(embedding.data.toString())
        bind(limit)
      }) { Document(string()) }
    }
}
